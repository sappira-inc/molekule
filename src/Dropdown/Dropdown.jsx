import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Portal } from 'react-portal';
import Popper from 'popper.js';
import Box from '../Box';

const DropdownTrigger = styled.div`
  display: inline-block;
`;

const DropdownMenu = styled.div`
  z-index: 10;
  position: absolute;
  background: white;
  border-radius: 4px;
  border: 1px solid #e4edf5;
  box-shadow: 0 0 3px 0 rgba(178, 194, 212, 0.3);
  min-width: 150px;
  width: ${p => `${p.width || 150}px`};
`;

export default class Dropdown extends React.Component {
  static propTypes = {
    trigger: PropTypes.element,
    render: PropTypes.func,
    autoclose: PropTypes.bool,
    placement: PropTypes.string,
    boundariesElement: PropTypes.string,
    on: PropTypes.string,
  };

  static defaultProps = {
    autoclose: true,
    placement: 'bottom-start',
    boundariesElement: 'window',
    on: 'click',
  };

  triggerRef = React.createRef();
  menuRef = React.createRef();

  state = {
    isOpen: false,
  };

  componentDidMount() {
    this.attachAutocloseListener();
    this.triggerRef.current.addEventListener(this.props.on, this.toggle);
  }

  componentWillUnmount() {
    if (this.positioner) {
      this.positioner.destroy();
    }
    this.detachAutocloseListener();
    this.triggerRef.current.removeEventListener(this.props.on, this.toggle);
  }

  componentDidUpdate() {
    if (this.positioner && this.positioner.update) {
      this.positioner.update();
    }
  }

  show = () => {
    const { placement, boundariesElement } = this.props;

    this.setState(
      {
        isOpen: true,
      },
      () => {
        this.positioner = new Popper(this.triggerRef.current, this.menuRef.current, {
          placement,
          modifiers: {
            offset: {
              offset: '0, 10',
            },
            flip: {
              behavior: ['left', 'bottom', 'top', 'right'],
            },
            preventOverflow: {
              boundariesElement,
            },
          },
        });
      }
    );
  };

  hide = () => {
    this.setState(
      {
        isOpen: false,
      },
      () => {
        this.positioner.destroy();
      }
    );
  };

  toggle = e => {
    e.stopPropagation();

    if (this.state.isOpen) {
      this.hide();
    } else {
      this.show();
    }
  };

  attachAutocloseListener() {
    if (!this.props.autoclose) {
      return;
    }
    document.addEventListener('click', this.autoclose);
  }

  detachAutocloseListener() {
    document.removeEventListener('click', this.autoclose);
  }

  autoclose = e => {
    if (this.state.isOpen && !this.menuRef.current.contains(e.target)) {
      this.hide();
    }
  };

  render() {
    const { width, trigger, render, children } = this.props;
    const { isOpen } = this.state;

    const renderFn = render || children;

    return (
      <Fragment>
        <DropdownTrigger innerRef={this.triggerRef} aria-haspopup="true" aria-expanded={isOpen}>
          {trigger}
        </DropdownTrigger>

        <Portal>
          {isOpen && (
            <DropdownMenu innerRef={this.menuRef} width={width}>
              {renderFn({
                close: this.toggle,
              })}
            </DropdownMenu>
          )}
        </Portal>
      </Fragment>
    );
  }
}

Dropdown.Header = styled.div`
  padding: 8px 12px;
  font-size: 1rem;
  border-bottom: 1px solid ${p => p.theme.colors.grayLight};
`;

Dropdown.Body = styled(Box)`
  padding: 12px;
`;

Dropdown.Title = styled.h3`
  display: block;
  font-weight: 600;
  font-size: 14px;
  color: ${p => p.theme.colors.primary};
  margin: 0 0 4px;
`;

Dropdown.Item = styled.div`
  opacity: ${p => (p.disabled ? 0.5 : 1)};
  pointer-events: ${p => (p.disabled ? 'none' : 'initial')};
  text-decoration: none;
  color: inherit;
  display: flex;
  align-items: center;
  margin-left: -12px;
  margin-right: -12px;
  padding: 4px 12px;
  font-size: 14px;
  cursor: pointer;

  & + ${Dropdown.Title} {
    margin-top: 12px;
  }

  &:hover {
    color: inherit;
  }
`;

Dropdown.Footer = styled.footer`
  background: {p => p.theme.colors.grayLightest};
  padding: 12px;
  border-radius: 0 0 4px 4px;
  border-top: 1px solid ${p => p.theme.colors.grayLight};
  font-size: 14px;
`;
