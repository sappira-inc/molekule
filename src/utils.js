import { get, kebabCase } from 'lodash';
import styled from 'styled-components';

export const themeGet = (theme, lookup, fallback) => get(theme, lookup, fallback);

export const getComponentVariant = (theme, componentName, variant) => {
  const config = themeGet(
    theme,
    `components.${componentName}.variants.${variant}`,
    theme.variants[componentName][variant]
  );
  if (!config) throw new Error(`Molekule: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentStyle = componentName => ({ theme }) => themeGet(theme, `components.${componentName}.style`);

export const getComponentClassName = ({ className, theme: { classPrefix }, variant }, name) =>
  `${className || ''} ${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

export const createComponent = ({ name, tag = 'div', as, style, props: defaultProps = () => ({}) }) => {
  const component = as ? styled(as) : styled[tag];

  return component.attrs(props => ({
    ...defaultProps(props),
    ...props,
    className: getComponentClassName(props, kebabCase(name)),
  }))`
    ${style}
    ${getComponentStyle(name)}
    ${({ styles = {} }) => styles[name] || {}}
  `;
};
