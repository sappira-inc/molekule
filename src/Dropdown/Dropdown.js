import React, { useRef, useState, useEffect, useContext } from 'react';
import PropTypes from 'prop-types';
import { FocusOn } from 'react-focus-on';
import { Transition } from 'react-transition-group';
import { Manager, Reference, Popper } from 'react-popper';
import { css, keyframes } from 'styled-components';
import Box from '../Box';
import Portal from '../Portal';
import Icon from '../Icon';
import { useKeyPress } from '../hooks';
import { createComponent, findNextFocusableElement, findPreviousFocusableElement } from '../utils';

const DropdownContext = React.createContext({});

export const PLACEMENT_TRANSITION_ORIGINS = {
  'top-start': '0 100%',
  top: '50% 100%',
  'top-end': '100% 100%',
  'right-start': '0 0',
  right: '0 50%',
  'right-end': '0 100%',
  'bottom-start': '0 0',
  bottom: '50% 0',
  'bottom-end': '100% 0',
  'left-start': '100% 0',
  left: '100% 50%',
  'left-end': '100% 100%',
};

const ARROW_KEYS = ['ArrowUp', 'ArrowDown'];

const Trigger = createComponent({
  name: 'DropdownTrigger',
  style: css`
    align-self: flex-start;
    display: inline-block;
    outline: none;
  `,
});

/** Easily display contextual overlays using custom trigger elements. Dropdowns positioning system uses [Popper.js](https://github.com/FezVrasta/popper.js). Refer to their documentation for placement and option overrides. */
export default function Dropdown({
  autoclose,
  placement,
  positionFixed,
  boundariesElement,
  offset,
  trigger,
  render,
  children,
  portalNode,
  styles = {},
  ...menuProps
}) {
  const popperRef = useRef();
  const menuRef = useRef();
  const triggerRef = useRef();

  const [isOpen, setOpen] = useState(false);

  const open = () => setOpen(true);
  const close = () => setOpen(false);
  const toggle = () => (isOpen ? close() : open());

  const handleTrigger = e => {
    // Allow all clicks and, for non-button elements, Enter and Space to toggle Dropdown
    // https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role#Required_JavaScript_Features
    if (e.type === 'click' || (e.type === 'keypress' && (e.which === 13 || e.which === 32))) {
      e.preventDefault();
      e.stopPropagation();
      toggle();
    }
  };

  // Wait for next tick after open to prevent page jump when focusing
  useEffect(() => {
    if (isOpen) {
      setTimeout(() => {
        if (menuRef.current) menuRef.current.focus();
      });
    }
  }, [isOpen]);

  useKeyPress('Escape', () => {
    if (isOpen) {
      if (triggerRef.current) triggerRef.current.focus();
      close();
    }
  });

  useKeyPress(ARROW_KEYS, event => {
    if (isOpen && menuRef.current) {
      event.preventDefault();
      const focusArgs = [menuRef.current, document.activeElement];
      const nextFocusable =
        event.key === 'ArrowUp' ? findPreviousFocusableElement(...focusArgs) : findNextFocusableElement(...focusArgs);

      if (nextFocusable) {
        nextFocusable.focus();
      }
    }
  });

  const handleMenuBlur = e => {
    if (autoclose && menuRef.current && !menuRef.current.contains(e.relatedTarget)) {
      close();
    }
  };

  const renderer = render || children;

  return (
    <DropdownContext.Provider value={{ close }}>
      <Manager>
        <Reference>
          {({ ref: passedTriggerRef }) => (
            <Trigger style={styles.Trigger} ref={passedTriggerRef} tabIndex={-1}>
              {React.cloneElement(trigger, {
                role: trigger.role || 'button',
                tabIndex: trigger.tabIndex || 0,
                'aria-haspopup': true,
                'aria-expanded': isOpen,
                onClick: handleTrigger,
                onKeyPress: handleTrigger,
                ref: triggerRef,
                style: {
                  cursor: 'pointer',
                  ...(trigger.style || {}),
                },
              })}
            </Trigger>
          )}
        </Reference>

        {isOpen && (
          <Portal node={portalNode}>
            <Popper
              innerRef={popperRef}
              placement={placement}
              positionFixed={positionFixed}
              modifiers={{
                offset: {
                  offset,
                },
                flip: {
                  enabled: false,
                },
                preventOverflow: {
                  boundariesElement,
                  padding: 12,
                },
                // Prevents wacky jumps in Transition component
                computeStyle: { gpuAcceleration: false },
              }}>
              {({ ref, style }) => (
                <Transition in={isOpen} timeout={0} appear>
                  {state => (
                    <FocusOn enabled={isOpen} autoFocus={false}>
                      <DropdownMenu
                        ref={menuInner => {
                          menuRef.current = menuInner;
                          ref(menuInner);
                        }}
                        style={style}
                        placement={placement}
                        transitionState={state}
                        onBlur={handleMenuBlur}
                        data-testid="dropdown-menu"
                        {...menuProps}>
                        {typeof renderer === 'function'
                          ? renderer({
                              close,
                            })
                          : renderer}
                      </DropdownMenu>
                    </FocusOn>
                  )}
                </Transition>
              )}
            </Popper>
          </Portal>
        )}
      </Manager>
    </DropdownContext.Provider>
  );
}

Dropdown.propTypes = {
  placement: PropTypes.oneOf(Object.keys(PLACEMENT_TRANSITION_ORIGINS)),
  trigger: PropTypes.element.isRequired,
  render: PropTypes.func,
  autoclose: PropTypes.bool,
  offset: PropTypes.string,
  boundariesElement: PropTypes.string,
  positionFixed: PropTypes.bool,
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  zIndex: PropTypes.number,
  transitionDuration: PropTypes.number,
  transitionTimingFunction: PropTypes.string,
  portalNode: PropTypes.instanceOf(typeof Element !== 'undefined' ? Element : Object),
};

Dropdown.defaultProps = {
  placement: 'bottom-start',
  autoclose: true,
  offset: '0, 10',
  positionFixed: false,
  boundariesElement: 'viewport',
  width: 'auto',
  zIndex: 10,
  transitionDuration: 225,
  transitionTimingFunction: 'cubic-bezier(0.25, 0.1, 0.17, 1.2)',
};

const scaleIn = keyframes`
  from {
    opacity: 0.75;
    transform: scale(0.75);
  }
  to {
    opacity: 1;
    transform: scale(1);
  }
`;

const DropdownMenu = createComponent({
  name: 'DropdownMenu',
  props: () => ({
    tabIndex: 0,
  }),
  style: ({ width, theme, placement, zIndex, transitionState, transitionDuration, transitionTimingFunction }) => css`
    z-index: ${zIndex};
    background: white;
    border-radius: ${theme.radius}px;
    border: 1px solid ${theme.colors.greyLighter};
    outline: none;
    box-shadow: ${theme.shadow.soft};
    width: ${typeof width === 'string' ? width : `${width}px`};
    opacity: 0.75;
    transform: scale(0.75);
    transform-origin: ${PLACEMENT_TRANSITION_ORIGINS[placement]};
    padding: 8px;

    ${(transitionState === 'entering' || transitionState === 'entered') &&
      css`
        animation: ${scaleIn} ${transitionDuration}ms ${transitionTimingFunction} forwards;
      `}
  `,
});

Dropdown.Divider = createComponent({
  name: 'DropdownDivider',
  tag: 'hr',
  style: ({ theme }) => css`
    background: ${theme.colors.greyLight};
    height: 1px;
    border: 0;
    margin-left: -8px;
    margin-right: -8px;
  `,
});

Dropdown.Title = createComponent({
  name: 'DropdownTitle',
  tag: 'span',
  style: ({ theme }) => css`
    display: block;
    font-weight: 700;
    color: ${theme.colors.greyDark};
  `,
});

const StyledDropdownItem = createComponent({
  name: 'DropdownItem',
  props: ({ disabled }) => ({
    tabIndex: disabled ? -1 : 0,
    role: 'button',
  }),
  as: Box,
  style: ({ disabled, theme, color, icon, iconProps, selected }) => css`
    display: block;
    flex: 1;
    opacity: ${disabled ? 0.3 : 1};
    pointer-events: ${disabled ? 'none' : 'initial'};
    user-select: ${disabled ? 'none' : 'initial'};
    text-decoration: none;
    color: ${color || theme.colors.greyDarkest};
    cursor: pointer;
    margin: 0;
    padding: 8px;
    transition: 125ms background;
    outline: none;
    appearance: none;
    border: 0;
    border-radius: ${theme.radius}px;
    font: inherit;
    font-size: 14px;
    font-weight: 500;
    text-align: left;
    position: relative;

    &:hover,
    &:focus {
      color: ${color || theme.colors.greyDarkest};
      background: ${theme.colors.greyLightest};
    }

    ${icon &&
      css`
        padding-left: ${(iconProps.size || 16) + 16}px;
      `}

    ${selected &&
      css`
        color: ${theme.colors.primary};

        &:hover,
        &:active,
        &:focus {
          color: ${theme.colors.primary};
        }
      `}
  `,
});

const StyledIcon = createComponent({
  name: 'DropdownIcon',
  as: Icon,
  style: css`
    position: absolute;
    left: 8px;
  `,
});

Dropdown.Item = ({ closeOnClick = true, onClick, children, icon, iconProps = {}, ...props }) => {
  const { close } = useContext(DropdownContext);
  const handleClick = () => {
    if (closeOnClick) {
      close();
    }
    if (typeof onClick === 'function') {
      onClick();
    }
  };
  return (
    <StyledDropdownItem onClick={handleClick} icon={icon} iconProps={iconProps} {...props}>
      {icon && <StyledIcon name={icon} {...iconProps} />}
      {children}
    </StyledDropdownItem>
  );
};
