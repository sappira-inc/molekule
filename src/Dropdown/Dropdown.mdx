---
menu: Core
name: Dropdown
---

import { useState } from 'react';
import { Playground, PropsTable } from 'docz';
import Dropdown, { PLACEMENT_TRANSITION_ORIGINS } from './Dropdown';
import Icon from '../Icon';
import Flex from '../Flex';
import RadioGroup from '../Form/RadioGroup';
import Button from '../Button';

# Dropdown

Easily display contextual overlays using custom trigger elements. Dropdown's positioning system uses [Popper.js](https://github.com/FezVrasta/popper.js). Refer to their documentation for placement and option overrides.

## Props
<PropsTable of={Dropdown} />

## Examples

<Playground>
  {() => {
    function Example() {
      const [placement, setPlacement] = useState('top');
      return (
        <>
          <Button.Group>
            <RadioGroup
              label={<strong>Placement</strong>}
              value={placement}
              choices={Object.keys(PLACEMENT_TRANSITION_ORIGINS).map(placement => ({
                value: placement,
                label: placement,
              }))}
              onChange={(_, val) => setPlacement(val)}
            />
            <Dropdown
              placement={placement}
              width={250}
              trigger={<Button variant="success">Open Dropdown</Button>}
            >
              <Dropdown.Header title="Dropdown" />

              <Dropdown.Body>
                <Dropdown.SectionTitle>Section One</Dropdown.SectionTitle>
                <Dropdown.Item closeOnClick={false}>I don't close when clicked</Dropdown.Item>
                <Dropdown.Item as="button">Item Two</Dropdown.Item>

                <Dropdown.SectionTitle>Section Two</Dropdown.SectionTitle>
                <Dropdown.Item as="a" onClick={close} href="http://google.com" target="_blank">Item One</Dropdown.Item>
                <Dropdown.Item disabled>Item Two</Dropdown.Item>
              </Dropdown.Body>

              <Dropdown.Footer>Footer</Dropdown.Footer>
            </Dropdown>

            <Dropdown
              width={250}
              trigger={<Button>Open Other Dropdown</Button>}
            >

              <Dropdown.Header title="Dropdown" />

              <Dropdown.Body>
                <Dropdown.SectionTitle>Section One</Dropdown.SectionTitle>
                <Dropdown.Item>Item One</Dropdown.Item>
                <Dropdown.Item>Item Two</Dropdown.Item>

                <Dropdown.SectionTitle>Section Two</Dropdown.SectionTitle>
                <Dropdown.Item>Item One</Dropdown.Item>
                <Dropdown.Item disabled>Item Two</Dropdown.Item>
              </Dropdown.Body>

              <Dropdown.Footer>Footer</Dropdown.Footer>
            </Dropdown>

            <div>
              <Dropdown
                width={250}
                placement="top"
                trigger={<Icon name="information-outline" />}
              >
                <Dropdown.Header title="Dropdown" />

                <Dropdown.Body>
                  <Dropdown.SectionTitle>Section One</Dropdown.SectionTitle>
                  <Dropdown.Item>Item One</Dropdown.Item>
                  <Dropdown.Item>Item Two</Dropdown.Item>
                </Dropdown.Body>

                <Dropdown.Footer>Footer</Dropdown.Footer>
              </Dropdown>
            </div>

            <Flex flex={1}>
              <Dropdown
                width={250}
                placement="top"
                styles={{ Trigger: { display: 'flex', flex: 1 } }}
                trigger={(<Flex flex={1} justifyContent="space-between">
                  <Button style={{ width: `100%`}}>Open Flex Dropdown</Button>
                </Flex>)}>
                <Dropdown.Header title="Dropdown" />

                <Dropdown.Body>
                  <Dropdown.SectionTitle>Section One</Dropdown.SectionTitle>
                  <Dropdown.Item>Item One</Dropdown.Item>
                  <Dropdown.Item>Item Two</Dropdown.Item>
                </Dropdown.Body>

                <Dropdown.Footer>Footer</Dropdown.Footer>
              </Dropdown>
            </Flex>
          </Button.Group>
        </>
      );
    }

    return <Example />
  }}

</Playground>
