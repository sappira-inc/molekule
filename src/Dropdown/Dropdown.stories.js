import React, { useState } from 'react';
import Dropdown, { PLACEMENT_TRANSITION_ORIGINS } from './Dropdown';
import Flex from '../Flex';
import RadioGroup from '../Form/RadioGroup';
import Button from '../Button';

export default {
  title: 'Components|Dropdown',
  component: Dropdown,
};

function BasicExample() {
  const [placement, setPlacement] = useState('bottom-start');
  return (
    <Flex justifyContent="center">
      <Flex mr={5}>
        <RadioGroup
          label={<strong>Placement</strong>}
          value={placement}
          choices={Object.keys(PLACEMENT_TRANSITION_ORIGINS).map(p => ({
            value: p,
            label: p,
          }))}
          onChange={(_, val) => setPlacement(val)}
        />
      </Flex>

      <Flex alignSelf="center">
        <Dropdown
          placement={placement}
          width={250}
          trigger={
            <Button mr={3} variant="primary">
              Basic Dropdown
            </Button>
          }>
          <Dropdown.Item>Dropdown Item</Dropdown.Item>
          <Dropdown.Item>Dropdown Item</Dropdown.Item>
          <Dropdown.Item>Dropdown Item</Dropdown.Item>
          <Dropdown.Item>Dropdown Item</Dropdown.Item>
          <Dropdown.Item color="red">Cancel</Dropdown.Item>
        </Dropdown>
      </Flex>
    </Flex>
  );
}

export const Basic = () => <BasicExample />;

export const WithTitles = () => (
  <Flex>
    <Dropdown placement="bottom-start" width={250} trigger={<Button variant="danger">Dropdown w/Titles</Button>}>
      <Dropdown.Title>Section Title</Dropdown.Title>
      <Dropdown.Item selected closeOnClick={false}>
        Dropdown Item
      </Dropdown.Item>
      <Dropdown.Item>Dropdown Item</Dropdown.Item>
      <Dropdown.Item>Dropdown Item</Dropdown.Item>
      <Dropdown.Divider />
      <Dropdown.Title>Section Title</Dropdown.Title>
      <Dropdown.Item>Dropdown Item</Dropdown.Item>
      <Dropdown.Item>Dropdown Item</Dropdown.Item>
      <Dropdown.Item>Dropdown Item</Dropdown.Item>
      <Dropdown.Divider />
      <Dropdown.Item color="red">Cancel</Dropdown.Item>
    </Dropdown>
  </Flex>
);

export const WithIcons = () => (
  <Flex>
    <Dropdown
      placement="bottom-start"
      width={250}
      trigger={
        <Button mr={3} variant="success">
          Dropdown w/Icons
        </Button>
      }>
      <Dropdown.Item icon="account-circle">Dropdown Item</Dropdown.Item>
      <Dropdown.Item icon="pencil">Dropdown Item</Dropdown.Item>
      <Dropdown.Divider />
      <Dropdown.Item icon="stethoscope">Dropdown Item</Dropdown.Item>
      <Dropdown.Item icon="bell">Dropdown Item</Dropdown.Item>
      <Dropdown.Divider />
      <Dropdown.Item icon="settings">Dropdown Item</Dropdown.Item>
      <Dropdown.Divider />
      <Dropdown.Item icon="trash-can" color="red">
        Cancel
      </Dropdown.Item>
    </Dropdown>
  </Flex>
);
