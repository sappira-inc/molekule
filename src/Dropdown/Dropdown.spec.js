import React from 'react';
import { renderWithTheme, fireEvent, wait } from '../../test/utils';
import Dropdown from './Dropdown';

jest.mock('popper.js', () => {
  const PopperJS = jest.requireActual('popper.js');

  return class Popper {
    static placements = PopperJS.placements;

    constructor() {
      return {
        destroy: () => {},
        scheduleUpdate: () => {},
      };
    }
  };
});

describe('<Dropdown />', () => {
  let renderUtils;

  const renderDropdown = (props = {}) => {
    const wrapper = document.createElement('div');
    wrapper.setAttribute('tabindex', 1);
    const utils = renderWithTheme(
      <Dropdown {...props} portalNode={wrapper} trigger={<div>Trigger</div>}>
        <Dropdown.Header>Header</Dropdown.Header>
        <Dropdown.Body>
          <Dropdown.Item data-testid="item-one">One</Dropdown.Item>
          <Dropdown.Item data-testid="item-two">Two</Dropdown.Item>
        </Dropdown.Body>
        <Dropdown.Footer>Footer</Dropdown.Footer>
      </Dropdown>,
      {
        container: document.body.appendChild(wrapper),
      }
    );
    return {
      wrapper,
      ...utils,
    };
  };

  const assertDropdownOpen = (utils = renderUtils) =>
    wait(() => {
      expect(utils.queryByText('Header')).toBeInTheDocument();
    });

  const assertDropdownClosed = (utils = renderUtils) =>
    wait(() => {
      expect(utils.queryByText('Header')).not.toBeInTheDocument();
    });

  const openDropdown = async (utils = renderUtils) => {
    const trigger = utils.getByText('Trigger');
    fireEvent.click(trigger, { stopPropagation: () => null });
    return assertDropdownOpen(utils);
  };

  beforeEach(() => {
    renderUtils = renderDropdown();
  });

  test('only renders trigger on mount', () => {
    expect(renderUtils.asFragment()).toMatchSnapshot();
  });

  // TODO: add cases for space and enter keypress events
  test('opens menu with focus when trigger is clicked', async () => {
    await openDropdown();
    expect(renderUtils.asFragment()).toMatchSnapshot();
    expect(renderUtils.getByTestId('dropdown-menu') === document.activeElement).toBeTruthy();
  });

  test('closes when escape is pressed', async () => {
    await openDropdown();
    fireEvent.keyDown(document.body, { key: 'Escape' });
    await assertDropdownClosed();
  });

  test('closes when menu loses focus', async () => {
    await openDropdown();

    // Some issues with fireEvent.focus: https://github.com/kentcdodds/react-testing-library/issues/276#issuecomment-473392827
    renderUtils.wrapper.focus();

    await assertDropdownClosed();
  });

  test('arrow keys navigate to focusable elements', async () => {
    await openDropdown();
    const itemOne = renderUtils.getByTestId('item-one');
    const itemTwo = renderUtils.getByTestId('item-two');
    const isFocused = node => node === document.activeElement;

    // First focusable element in tree should be selected on first arrow down
    fireEvent.keyDown(document.body, { key: 'ArrowDown' });
    expect(isFocused(itemOne)).toBeTruthy();

    // Arrow up on first item should bring us to last item
    fireEvent.keyDown(document.body, { key: 'ArrowUp' });
    expect(isFocused(itemTwo)).toBeTruthy();

    fireEvent.keyDown(document.body, { key: 'ArrowUp' });
    expect(isFocused(itemOne)).toBeTruthy();
  });

  describe('prop: width', () => {
    test('wdith defaults to auto', async () => {
      const utils = renderDropdown();
      await openDropdown(utils);
      expect(utils.getByTestId('dropdown-menu')).toHaveStyleRule('width', 'auto');
    });

    test('supports string widths', async () => {
      const utils = renderDropdown({ width: '2rem' });
      await openDropdown(utils);
      expect(utils.getByTestId('dropdown-menu')).toHaveStyleRule('width', '2rem');
    });

    test('supports number widths', async () => {
      const utils = renderDropdown({ width: 200 });
      await openDropdown(utils);
      expect(utils.getByTestId('dropdown-menu')).toHaveStyleRule('width', '200px');
    });
  });
});
