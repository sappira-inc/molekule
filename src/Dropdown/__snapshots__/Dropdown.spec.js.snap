// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> only renders trigger on mount 1`] = `
<DocumentFragment>
  .c0 {
  font-family: inherit;
  display: inline-block;
  text-align: center;
  cursor: pointer;
  text-transform: capitalize;
  font-weight: bold;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border-radius: 4px;
  pointer-events: auto;
  opacity: 1;
  color: white;
  height: 36px;
  padding: 0 18px;
  font-size: 12px;
  width: auto;
  background: #2DAAF2;
  border: 1px solid #2DAAF2;
  -webkit-transition: 175ms;
  transition: 175ms;
}

.c0:hover {
  background: #45b4f3;
  border-color: #45b4f3;
}

.c0:active {
  background: #51b9f4;
  border-color: #51b9f4;
}

<button
    aria-expanded="false"
    aria-haspopup="true"
    class="re-button re-button-primary c0"
  >
    Trigger
  </button>
</DocumentFragment>
`;

exports[`<Dropdown /> opens menu with focus when trigger is clicked 1`] = `
<DocumentFragment>
  .c0 {
  font-family: inherit;
  display: inline-block;
  text-align: center;
  cursor: pointer;
  text-transform: capitalize;
  font-weight: bold;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border-radius: 4px;
  pointer-events: auto;
  opacity: 1;
  color: white;
  height: 36px;
  padding: 0 18px;
  font-size: 12px;
  width: auto;
  background: #2DAAF2;
  border: 1px solid #2DAAF2;
  -webkit-transition: 175ms;
  transition: 175ms;
}

.c0:hover {
  background: #45b4f3;
  border-color: #45b4f3;
}

.c0:active {
  background: #51b9f4;
  border-color: #51b9f4;
}

<button
    aria-expanded="true"
    aria-haspopup="true"
    class="re-button re-button-primary c0"
  >
    Trigger
  </button>
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="1"
  />
  .c0 {
  z-index: 10;
  background: white;
  border-radius: 4px;
  border: 1px solid #e4edf5;
  outline: none;
  box-shadow: 0 0 3px 0 rgba(178,194,212,0.3);
  width: 150px;
  opacity: 0.75;
  -webkit-transform: scale(0.75);
  -ms-transform: scale(0.75);
  transform: scale(0.75);
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
  -webkit-animation: gPsOjp 225ms cubic-bezier(0.25,0.1,0.17,1.2) forwards;
  animation: gPsOjp 225ms cubic-bezier(0.25,0.1,0.17,1.2) forwards;
}

.c1 {
  padding: 0.75rem 1rem 0;
}

.c2 {
  padding: 0 0 0.25rem;
  border-bottom: 2px solid #DEE0E4;
}

.c3 {
  padding: 1rem;
}

.c4 {
  display: block;
  width: calc(100% + 2rem);
  opacity: 1;
  pointer-events: initial;
  -webkit-user-select: initial;
  -moz-user-select: initial;
  -ms-user-select: initial;
  user-select: initial;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  cursor: pointer;
  margin: 0 calc(-1rem);
  padding: 0.25rem 1rem;
  -webkit-transition: 125ms background;
  transition: 125ms background;
  outline: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 0;
  font: inherit;
  text-align: left;
}

.c4 + .c6 {
  margin-top: 1rem;
}

.c4:hover,
.c4:focus {
  color: inherit;
  background: #F1F4F6;
}

.c5 {
  background: #F1F4F6;
  padding: 0.75rem 1rem;
  border-radius: 0 0 4px 4px;
}

<div
    data-focus-lock-disabled="false"
  >
    <div
      class="re-dropdown-menu c0"
      data-testid="dropdown-menu"
      style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
      tabindex="0"
      width="150"
    >
      <header
        class="re-dropdown-header c1"
      >
        <div
          class="re-dropdown-header-inner c2"
        >
          Header
        </div>
      </header>
      <div
        class="re-box re-dropdown-body c3"
      >
        <div
          class="re-box re-dropdown-item c4"
          data-testid="item-one"
          role="button"
          tabindex="0"
        >
          One
        </div>
        <div
          class="re-box re-dropdown-item c4"
          data-testid="item-two"
          role="button"
          tabindex="0"
        >
          Two
        </div>
      </div>
      <footer
        class="re-box re-dropdown-footer c5"
      >
        Footer
      </footer>
    </div>
  </div>
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />
</DocumentFragment>
`;
